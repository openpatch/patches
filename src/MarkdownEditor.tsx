import { css } from "@emotion/react";
import { useRef, useState } from "react";
import { Box } from "./Box";
import { ButtonSecondary } from "./ButtonSecondary";
import { CodeEditor, CodeEditorProps, CodeEditorRef } from "./CodeEditor";
import { Markdown } from "./Markdown";

export type MarkdownEditorProps = {
  allowPreview?: boolean;
  locale?: {
    preview: string;
    edit: string;
  };
} & Omit<CodeEditorProps, "language">;

const defaultLocale = {
  preview: "Preview",
  edit: "Edit",
};

export const MarkdownEditor = ({
  lineNumbers = false,
  value = "",
  onChange = () => {},
  variant,
  allowPreview = false,
  locale = defaultLocale,
  height,
  ...props
}: MarkdownEditorProps) => {
  const editor = useRef<CodeEditorRef>();
  const [isPreview, setIsPreview] = useState(false);

  function getContent() {
    return editor.current?.state.doc.toString() || "";
  }

  return (
    <Box
      position="relative"
      minHeight="80px"
      height={height}
      overflowY="auto"
      css={(theme) => [
        variant === "outlined" &&
          css`
            border-style: solid;
            border-radius: ${theme.radii.small};
            border-width: ${theme.borderWidths.standard};
            border-color: ${theme.colors.neutral["200"]};

            > div,
            > div > div {
              border-radius: ${theme.radii.small};
            }
          `,
        variant === "input" &&
          css`
            border-radius: ${theme.radii.small};
            border-color: ${theme.colors.neutral["100"]};
            border-style: solid;
            border-width: ${theme.borderWidths.standard};

            > div,
            > div > div {
              border-radius: ${theme.radii.small};
            }
          `,
      ]}
    >
      {allowPreview && (
        <ButtonSecondary
          tone="accent"
          onClick={() => setIsPreview((p) => !p)}
          css={(theme) => css`
            position: absolute;
            top: 5px;
            right: 5px;
            font-size: ${theme.fontSizes.xsmall};
            z-index: 99;
            opacity: 0.8;
          `}
        >
          {isPreview ? locale.edit : locale.preview}
        </ButtonSecondary>
      )}
      {isPreview ? (
        <Box p="xxsmall">
          <Markdown markdown={getContent()} />
        </Box>
      ) : (
        <CodeEditor
          ref={editor}
          variant="normal"
          language="markdown"
          wrapLines
          lineNumbers={lineNumbers}
          value={value}
          onChange={onChange}
          {...props}
        />
      )}
    </Box>
  );
};
