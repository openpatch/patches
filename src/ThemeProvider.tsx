import React, { ReactElement } from "react";
import { ThemeProvider as EThemeProvider } from "emotion-theming";
import { Theme, base } from "./themes";
import { GlobalCssTheme } from "./themes/utils";

interface ThemeProviderProps {
  theme?: Theme;
  children: ReactElement;
}

const ThemeProvider = ({ theme = base, children }: ThemeProviderProps) => {
  return (
    <EThemeProvider theme={theme}>
      <GlobalCssTheme theme={theme} />
      {children}
    </EThemeProvider>
  );
};

export default ThemeProvider;
