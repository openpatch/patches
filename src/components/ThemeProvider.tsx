import React, { ReactElement } from "react";
import {
  ThemeProvider as SCThemeProvider,
  DefaultTheme,
} from "styled-components";
import defaultTheme from "../shared/theme";

interface ThemeProviderProps {
  theme?: DefaultTheme;
  children: ReactElement;
}

const ThemeProvider = ({
  theme = defaultTheme,
  children,
}: ThemeProviderProps) => {
  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};

export default ThemeProvider;
