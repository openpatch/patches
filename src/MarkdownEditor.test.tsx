import { render } from "@testing-library/react";
import { MarkdownEditor, MarkdownEditorProps } from "./MarkdownEditor";
import { ThemeProvider } from "./ThemeProvider";

const renderWithTheme = (props?: MarkdownEditorProps) =>
  render(
    <ThemeProvider>
      <MarkdownEditor {...props} />
    </ThemeProvider>
  );

test("should render", () => {
  renderWithTheme();
});

test("should match snapshot", () => {
  const { container } = renderWithTheme();
  expect(container).toMatchSnapshot();
});
