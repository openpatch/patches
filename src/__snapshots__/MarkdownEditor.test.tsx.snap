// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
.emotion-0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  position: relative;
  min-height: 80px;
  overflow-y: auto;
}

.emotion-2 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  height: auto;
  overflow-y: auto;
}

<div>
  <div
    id="patches-root"
  >
    <div
      class="emotion-0 emotion-1"
    >
      <div
        class="emotion-2 emotion-1"
        height="auto"
      >
        <div>
          <div
            class="cm-editor ͼ1 ͼ2 ͼ4"
          >
            <div
              aria-live="polite"
              style="position: absolute; top: -10000px;"
            />
            <div
              class="cm-scroller"
              tabindex="-1"
            >
              <div
                aria-multiline="true"
                autocapitalize="off"
                autocorrect="off"
                class="cm-content cm-lineWrapping"
                contenteditable="true"
                role="textbox"
                spellcheck="false"
                style="tab-size: 4"
                translate="no"
              >
                <div
                  class="cm-line"
                >
                  <br />
                </div>
              </div>
              <div
                aria-hidden="true"
                class="cm-selectionLayer"
              />
              <div
                aria-hidden="true"
                class="cm-cursorLayer"
                style="animation-duration: 1200ms;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
